src/Glpi/Asset/Asset.php

```diff
@@ -370,21 +382,99 @@ public function showForm($ID, array $options = [])
             [
                 'item'   => $this,
                 'params' => $options,
+                'custom_field_definitions' => static::getDefinition()->getCustomFieldDefinitions(),
             ]
         );
         return true;
     }
 
     public function prepareInputForAdd($input)
     {
+        $input = $this->handleCustomFieldsUpdate($input);
         return $this->prepareDefinitionInput($input);
     }
 
     public function prepareInputForUpdate($input)
     {
+        $input = $this->handleCustomFieldsUpdate($input);
         return $this->prepareDefinitionInput($input);
     }
 
+    protected function handleCustomFieldsUpdate(array $input): array
+    {
+        $custom_fields = [];
+
+        foreach (static::getDefinition()->getCustomFieldDefinitions() as $custom_field) {
+            if ($custom_field->fields['type'] === 'placeholder') {
+                continue;
+            }
+            $custom_field_name = 'custom_' . $custom_field->fields['name'];
+            if (!isset($input[$custom_field_name])) {
+                continue;
+            }
+            $value = $input[$custom_field_name];
+
+            if (!$custom_field->validateValue($value)) {
+                continue;
+            }
+
+            $custom_fields[$custom_field->getID()] = $value;
```

Comment:
In a discussion we had a month ago with @orthagh, he says that the custom fields JSON should use the user defined system name. For instance:
```diff
- {"1": "7595AB4"}
+ {"qa_number": "7595AB4"}
```

It may be a good idea if we consider that people may want to produce scripts based on the Rest API or based on DB extracts.

If we do this, then this system name should be read-only once it is defined. Changing it should be a specific action, that we could implement later. For the asset definition system name, we planned to develop it later, with the handling of a mapping from old system names to the current one, in order to be able to change it without breaking all scripts based on APIs. We should probably plan to do the same here.

Reply:
We can handle swapping the ID for the name at the application level.

I think BC is less of an issue here since the names don't trigger the generation of any PHP classes. We can add a warning that changing the name will break any existing API scripts that reference the field. Handling it ourselves seems like complexity that isn't needed in the code or in the UI for the users. Unless they specifically plan to use the field through the API, the user shouldn't care what it is. The way it is now, the name can be changed without having to bulk update every asset's JSON.

Reply:
> We can handle swapping the ID for the name at the application level.

In fact, it already is. For those who choose to use the legacy API, they will see their custom fields by name right next to the core fields with a "custom_" prefix and they can set the values for those fields the same way.